primitives/src/erasure_code/mod.rs

```diff
@@ -28,9 +30,17 @@ pub trait ErasureCode<T> {
         + CanonicalDeserialize;
 
     /// Encode `data` into `data.len() + parity_size` shares.
-    fn encode(data: &[T], parity_size: usize) -> Result<Vec<Self::Share>, PrimitivesError>;
+    fn encode<D>(data: D, parity_size: usize) -> Result<Vec<Self::Share>, PrimitivesError>
+    where
+        D: IntoIterator,
+        D::Item: Borrow<T>,
+        D::IntoIter: ExactSizeIterator + Clone;
 
```

Comment:
May I ask if we can simply accept an `Iterator` instead?

My thinking is as follows:
either we expect the caller of this `encode` to create an iterator and pass in that iter. (this is the example @ggutoski gave in the issue https://github.com/EspressoSystems/jellyfish/issues/235)
or we let the caller to pass in that super-struct `T` directly and expect `T: IntoIter`. 

```rust
// upstream
struct MyCollection { shares: T }
impl<T> IntoIter<Item: Borrow<T>> for MyCollection {}

decode(my_collection); 
// downstream
fn encode(data: D) where D: IntoIterator, D::Item: Borrow<T>
```

or 

```rust
// upstream
struct MyStruct { shares: T }
let my_struct_vec: Vec<MyStruct> = ...;

decode(my_struct_vec.iter().map(|d| &d.s)); 

// downstream
fn encode(data: I) where Iterator<Item: Borrow<T>>
```

Reply:
> May I ask if we can simply accept an `Iterator` instead?

From [stackoverflow](https://stackoverflow.com/questions/57543399/how-to-properly-pass-iterators-to-a-function-in-rust/57545430#57545430):

> Write functions that accept `impl IntoIterator<...>`. Because all Iterators implement `IntoIterator`, this is strictly more general than a function that accepts only `impl Iterator<...>`.

If we accept only `Iterator` the the caller is forced to give us an iterator. So, if the caller already has a collection `my_collection` of type `&[T]` (or any other collection type) then he can't just pass `my_collection` but must instead call `my_collection.iter()` first.

By accepting `IntoIterator` the function works regardless of whether we're given a collection or a custom iterator. We allow the use case described in #235 without changing the syntax for callers who just expect to give us a slice.

Not sure I understand your code snippets, but both use cases should work fine when `decode` accepts `IntoIterator`.

Reply:
yeah, agree that `IntoIterator` is a more relaxed/general API, which should be preferred.

Reply:
Note that tests that call `decode` did not even need to change to accommodate this new API.